RBAC Conflict after Traefik Installation

Hello, I appear to be having an issue with k8s_gateway (1.2.0) after installing both it and Traefik proxy (v3.4.0). I have found a workaround for now but wanted to raise an issue for visibility.

I've installed both using Helm, however currently am only watching `Ingress` and `Service` in the k8s_gateway configuration. I installed and configured k8s_gateway first, and had it working to resolve internal DNS names to both of the mentioned resources. However after installing Traefik and subsequently restarting k8s_gateway to make a minor configuration change k8s_gateway started throwing errors & has stopped working. The configuration change was unrelated AFAIK; I removed a `hosts` stanza from the `extraZonePlugins` section.

```
W0523 05:25:56.976365       1 reflector.go:569] pkg/mod/k8s.io/client-go@v0.32.4/tools/cache/reflector.go:251: failed to list *v1.Gateway: gateways.gateway.networking.k8s.io is forbidden: User "system:serviceaccount:internal-dns:coredns-ns1-k8s-gateway" cannot list resource "gateways" in API group "gateway.networking.k8s.io" at the cluster scope
E0523 05:25:56.976478       1 reflector.go:166] "Unhandled Error" err="pkg/mod/k8s.io/client-go@v0.32.4/tools/cache/reflector.go:251: Failed to watch *v1.Gateway: failed to list *v1.Gateway: gateways.gateway.networking.k8s.io is forbidden: User \"system:serviceaccount:internal-dns:coredns-ns1-k8s-gateway\" cannot list resource \"gateways\" in API group \"gateway.networking.k8s.io\" at the cluster scope" logger="UnhandledError"
...
[ERROR] plugin/errors: 2 nginx.ceto.internal. A: plugin/k8s_gateway: could not sync required resources
```

After Googling this looked to me like an RBAC issue, so I tracked it down to [this ClusterRole CRD](https://github.com/k8s-gateway/k8s_gateway/blob/master/charts/k8s-gateway/templates/rbac.yaml#L43-L51). I updated my configuration to include `HTTPRoute` in the `watchedResources` section to enable the permissions for "gateway.networking.k8s.io", upgraded the deployment through Helm and now k8s_gateway is working again (checking via `dig`).

So it looks to me like k8s_gateway is attempting to watch gateway resources even if they're not in the `watchedResources` list after the first installation of Traefik. I'm not really an expert so don't want to postulate too much more, though perhaps some resources created by the Traefik install have affected k8s_gateway?

Either way LMK if you want help to reproduce the issue, I can post the commands that I ran and my config if necessary.

--- a/kubernetes.go
+++ b/kubernetes.go
@@ -64,7 +64,18 @@ func newKubeController(ctx context.Context, c *kubernetes.Clientset, gw *gateway
 		gwClient: gw,
 	}
 
-	if crdExists(apiextensionsClient, "gatewayclasses.gateway.networking.k8s.io") {
+	configuredResources := dereferenceStrings(originalGateway.ConfiguredResources)
+	routingResources := []string{"HTTPRoute", "TLSRoute", "GRPCRoute"}
+
+	shouldInitGateway := false
+	for _, r := range routingResources {
+		if slices.Contains(configuredResources, r) {
+			shouldInitGateway = true
+			break
+		}
+	}
+
+	if crdExists(apiextensionsClient, "gatewayclasses.gateway.networking.k8s.io") && shouldInitGateway {
 		gatewayController := cache.NewSharedIndexInformer(
 			&cache.ListWatch{
 				ListFunc:  gatewayLister(ctx, ctrl.gwClient, core.NamespaceAll),
@@ -77,59 +88,61 @@ func newKubeController(ctx context.Context, c *kubernetes.Clientset, gw *gateway
 		ctrl.controllers = append(ctrl.controllers, gatewayController)
 		log.Infof("GatewayAPI controller initialized")
 
-		routingResources := []string{"HTTPRoute", "TLSRoute", "GRPCRoute"}
 		for _, resourceName := range routingResources {
-			if slices.Contains(dereferenceStrings(originalGateway.ConfiguredResources), resourceName) {
-				if resource := originalGateway.lookupResource(resourceName); resource != nil {
-					switch resourceName {
-					case "HTTPRoute":
-						httpRouteController := cache.NewSharedIndexInformer(
-							&cache.ListWatch{
-								ListFunc:  httpRouteLister(ctx, ctrl.gwClient, core.NamespaceAll),
-								WatchFunc: httpRouteWatcher(ctx, ctrl.gwClient, core.NamespaceAll),
-							},
-							&gatewayapi_v1.HTTPRoute{},
-							defaultResyncPeriod,
-							cache.Indexers{httpRouteHostnameIndex: httpRouteHostnameIndexFunc},
-						)
-						resource.lookup = lookupHttpRouteIndex(httpRouteController, gatewayController, originalGateway.resourceFilters.gatewayClasses)
-						ctrl.controllers = append(ctrl.controllers, httpRouteController)
-						log.Infof("HTTPRoute controller initialized")
+			if !slices.Contains(configuredResources, resourceName) {
+				continue
+			}
+			resource := originalGateway.lookupResource(resourceName)
+			if resource == nil {
+				continue
+			}
 
-					case "TLSRoute":
-						tlsRouteController := cache.NewSharedIndexInformer(
-							&cache.ListWatch{
-								ListFunc:  tlsRouteLister(ctx, ctrl.gwClient, core.NamespaceAll),
-								WatchFunc: tlsRouteWatcher(ctx, ctrl.gwClient, core.NamespaceAll),
-							},
-							&gatewayapi_v1alpha2.TLSRoute{},
-							defaultResyncPeriod,
-							cache.Indexers{tlsRouteHostnameIndex: tlsRouteHostnameIndexFunc},
-						)
-						resource.lookup = lookupTLSRouteIndex(tlsRouteController, gatewayController, originalGateway.resourceFilters.gatewayClasses)
-						ctrl.controllers = append(ctrl.controllers, tlsRouteController)
-						log.Infof("TLSRoute controller initialized")
+			switch resourceName {
+			case "HTTPRoute":
+				httpRouteController := cache.NewSharedIndexInformer(
+					&cache.ListWatch{
+						ListFunc:  httpRouteLister(ctx, ctrl.gwClient, core.NamespaceAll),
+						WatchFunc: httpRouteWatcher(ctx, ctrl.gwClient, core.NamespaceAll),
+					},
+					&gatewayapi_v1.HTTPRoute{},
+					defaultResyncPeriod,
+					cache.Indexers{httpRouteHostnameIndex: httpRouteHostnameIndexFunc},
+				)
+				resource.lookup = lookupHttpRouteIndex(httpRouteController, gatewayController, originalGateway.resourceFilters.gatewayClasses)
+				ctrl.controllers = append(ctrl.controllers, httpRouteController)
+				log.Infof("HTTPRoute controller initialized")
 
-					case "GRPCRoute":
-						grpcRouteController := cache.NewSharedIndexInformer(
-							&cache.ListWatch{
-								ListFunc:  grpcRouteLister(ctx, ctrl.gwClient, core.NamespaceAll),
-								WatchFunc: grpcRouteWatcher(ctx, ctrl.gwClient, core.NamespaceAll),
-							},
-							&gatewayapi_v1.GRPCRoute{},
-							defaultResyncPeriod,
-							cache.Indexers{grpcRouteHostnameIndex: grpcRouteHostnameIndexFunc},
-						)
-						resource.lookup = lookupGRPCRouteIndex(grpcRouteController, gatewayController, originalGateway.resourceFilters.gatewayClasses)
-						ctrl.controllers = append(ctrl.controllers, grpcRouteController)
-						log.Infof("GRPCRoute controller initialized")
-					}
-				}
+			case "TLSRoute":
+				tlsRouteController := cache.NewSharedIndexInformer(
+					&cache.ListWatch{
+						ListFunc:  tlsRouteLister(ctx, ctrl.gwClient, core.NamespaceAll),
+						WatchFunc: tlsRouteWatcher(ctx, ctrl.gwClient, core.NamespaceAll),
+					},
+					&gatewayapi_v1alpha2.TLSRoute{},
+					defaultResyncPeriod,
+					cache.Indexers{tlsRouteHostnameIndex: tlsRouteHostnameIndexFunc},
+				)
+				resource.lookup = lookupTLSRouteIndex(tlsRouteController, gatewayController, originalGateway.resourceFilters.gatewayClasses)
+				ctrl.controllers = append(ctrl.controllers, tlsRouteController)
+				log.Infof("TLSRoute controller initialized")
+
+			case "GRPCRoute":
+				grpcRouteController := cache.NewSharedIndexInformer(
+					&cache.ListWatch{
+						ListFunc:  grpcRouteLister(ctx, ctrl.gwClient, core.NamespaceAll),
+						WatchFunc: grpcRouteWatcher(ctx, ctrl.gwClient, core.NamespaceAll),
+					},
+					&gatewayapi_v1.GRPCRoute{},
+					defaultResyncPeriod,
+					cache.Indexers{grpcRouteHostnameIndex: grpcRouteHostnameIndexFunc},
+				)
+				resource.lookup = lookupGRPCRouteIndex(grpcRouteController, gatewayController, originalGateway.resourceFilters.gatewayClasses)
+				ctrl.controllers = append(ctrl.controllers, grpcRouteController)
+				log.Infof("GRPCRoute controller initialized")
 			}
 		}
 	}
 
-	// Handle Ingress and Service explicitly
 	for _, resourceName := range []string{"Ingress", "Service"} {
 		if slices.Contains(dereferenceStrings(originalGateway.ConfiguredResources), resourceName) {
 			if resource := originalGateway.lookupResource(resourceName); resource != nil {

